File: apps/mobile/modules/ricecal-widgets/ios/RiceCalWidgetsModule.swift
import ExpoModulesCore
import WidgetKit

/**
 The app's end of the App Group.

 Four calls, and every one of them is a few bytes into `UserDefaults`, so three
 of them are synchronous `Function`s rather than `AsyncFunction`s: the snapshot
 is written from a react-query subscription that fires whenever the day changes,
 and hopping to another queue and back would put the home screen a frame behind
 the diary for no benefit.

 `installedWidgetsAsync` is the exception because WidgetKit's own answer arrives
 on a completion handler.
 */
public class RiceCalWidgetsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("RiceCalWidgets")

    Function("setSnapshot") { (json: String) in
      RiceCalWidgetStore.writeSnapshot(json)
      RiceCalWidgetStore.reloadAll()
    }

    Function("clearSnapshot") {
      RiceCalWidgetStore.clearSnapshot()
      RiceCalWidgetStore.reloadAll()
    }

    /**
     A no-op, and deliberately not left out of the module.

     Android needs telling which scheme a widget tap should use, because a
     widget there lives in a library that cannot read the app's own intent
     filters. Here the config plugin has already written the same value into
     the extension's Info.plist, which is strictly better: a widget with no
     snapshot on it still has to be able to open the app, and that can happen
     before JS has ever run on this install.

     Kept so the two platforms present the same module and the JS seam does not
     have to ask which one it is talking to.
     */
    Function("setScheme") { (_: String) in
    }

    Function("takePendingActions") { () -> String in
      RiceCalWidgetStore.takePendingActions()
    }

    AsyncFunction("installedWidgetsAsync") { (promise: Promise) in
      WidgetCenter.shared.getCurrentConfigurations { result in
        switch result {
        case .success(let widgets):
          /**
           One entry per PLACED widget, so the same kind twice is two rows.
           Reduced to a set of kinds here rather than in JS: what the adoption
           report asks is which widgets somebody uses, and two copies of the
           water widget is one answer to that, not two.
           */
          let kinds = Set(widgets.compactMap { RiceCalWidgetKind(rawValue: $0.kind)?.reported })
          promise.resolve(Array(kinds).sorted())
        case .failure:
          // Not an error worth raising. The service is unavailable on a
          // simulator running an OS that never had WidgetKit, and on a device
          // where the extension has not been registered yet — neither is
          // something the app can act on, and the JS side reads an empty list
          // as "none installed", which is true in both cases.
          promise.resolve([String]())
        }
      }
    }
  }
}

File: apps/mobile/modules/ricecal-widgets/android/src/main/java/expo/modules/ricecalwidgets/RiceCalWidgetsModule.kt
package expo.modules.ricecalwidgets

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * The app's end of the widget store.
 *
 * The same five calls as the iOS module, and synchronous for the same reason:
 * each is a few bytes into `SharedPreferences`, written from a react-query
 * subscription that fires whenever the day changes.
 *
 * `installedWidgetsAsync` is async only to match the shape iOS needs, since
 * WidgetKit answers on a completion handler where `AppWidgetManager` answers
 * straight away.
 *
 * Every body goes through `withContext`, which is `?.let` rather than an early
 * `return@Function`: these lambdas are typed as returning a value, so a bare
 * return is a compile error rather than a no-op.
 */
class RiceCalWidgetsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("RiceCalWidgets")

    Function("setSnapshot") { json: String ->
      withContext {
        WidgetStore.writeSnapshot(it, json)
        WidgetStore.reloadAll(it)
      }
    }

    Function("clearSnapshot") {
      withContext {
        WidgetStore.clear(it)
        WidgetStore.reloadAll(it)
      }
    }

    /**
     * How a widget tap addresses the app.
     *
     * Told rather than discovered, because the scheme moves with the build
     * variant — a development client registers `ricecal-dev://` — and nothing
     * inside a library module can read the app's own intent filters. Until this
     * has been called once, a tap falls back to simply launching the app.
     *
     * A no-op on iOS: the config plugin writes the same value into the widget
     * extension's Info.plist at prebuild time, because a widget with no snapshot
     * still has to be able to open the app and JS may never have run.
     */
    Function("setScheme") { scheme: String ->
      withContext { WidgetStore.writeScheme(it, scheme) }
    }

    Function("takePendingActions") { ->
      context()?.let { WidgetStore.takePendingActions(it) } ?: "[]"
    }

    // The element type is spelt out because `emptyList()` on its own gives the
    // compiler nothing to infer from, and the error it produces names the type
    // parameter rather than the call.
    AsyncFunction("installedWidgetsAsync") { ->
      context()?.let { WidgetStore.installed(it) } ?: emptyList<String>()
    }
  }

  /**
   * The Android context, or nothing.
   *
   * Null while the module is being torn down, which is a real state rather than
   * a defensive habit: `WidgetSync` writes on a react-query subscription, and a
   * reload during development can fire one after the host has gone.
   */
  private fun context(): Context? = appContext.reactContext

  private inline fun withContext(body: (Context) -> Unit) {
    context()?.let(body)
  }
}
